AzureCalling/External/Authentication/AADAuthHandler.swift

```diff
@@ -53,163 +74,173 @@ class AADAuthHandler {
+                }
 
-            self?.acquireTokenSilently(from: viewController, account: account) {
-                completionHandler()
-            }
-        }
-    }
+                updateAccessToken(result?.accessToken)
+                currentAccount = result?.account
 
-    func acquireTokenInteractively(from viewController: UIViewController,
-                                   completionHandler: @escaping () -> Void) {
-        guard let applicationContext = self.applicationContext else {
-            return
-        }
-        let webViewParamaters = MSALWebviewParameters(authPresentationViewController: viewController)
-        webViewParamaters.webviewType = .wkWebView
+                let profile = try await getProfile()
+                userDisplayName = profile.displayName
 
-        let parameters = MSALInteractiveTokenParameters(scopes: appSettings.aadScopes, webviewParameters: webViewParamaters)
-        parameters.promptType = .selectAccount
-        parameters.completionBlockQueue = DispatchQueue.main
+                // Uncomment to obtain the user's avatar image if they have one
+//                let avatar = try? await getAvatarImage()
+//                userAvatar = avatar
 
-        applicationContext.acquireToken(with: parameters) { [weak self] (result, error) in
-
-            if let error = error {
-                print("Could not acquire token: \(error)")
-                completionHandler()
-                return
+                DispatchQueue.main.async {
+                    completionHandler(nil)
+                }
+            } catch {
+                DispatchQueue.main.async {
+                    completionHandler(error)
+                }
             }
+        }
+    }
 
-            guard let result = result else {
-                print("Could not acquire token: No result returned")
-                completionHandler()
-                return
+    func signOut(presentingVc viewController: UIViewController,
+                 completionHandler: @escaping (Error?) -> Void) {
+        Task {
+            do {
+                try await signOut(from: viewController)
+                print("MSAL sign out completed")
+                DispatchQueue.main.async {
+                    completionHandler(nil)
+                }
+            } catch {
+                print("MSAL couldn't sign out account with error: \(error)")
+                DispatchQueue.main.async {
+                    completionHandler(error)
+                }
             }
-
-            self?.updateAccessToken(result.accessToken)
-            self?.updateCurrentAccount(result.account)
-            completionHandler()
-
         }
     }
 
-    func signOutCurrentAccount(from viewController: UIViewController,
-                               completionHandler: @escaping () -> Void) {
-        guard let applicationContext = self.applicationContext,
+    // MARK: Private Functions
+    @MainActor
```

Comment:
Interesting choice marking the method as `MainActor`.  I haven't personally used this attribute, but is there a reason we would want to only have specific methods marked and not the whole class. I imagine all the service calls would require this?

Reply:
This _has_ to be this way as the calls to the MSAL library functions that touch the UIKit controls need to be on the main thread. Others do not have this restriction, so there's no need to force the main thread for tasks.